Content libraries settings overrides (SOL-46)

bradenmacdonald opened the pull request:
**Background**: This PR allows course authors to use blocks from a content library but override any `Scope.settings` fields such as `display_name` or `weight`. 
**Discussions**: Basic idea was presented on the wiki page for content libraries and discussed w/ a few people at edX
**Dependencies**: Depends on the feature branch https://github.com/edx/edx-platform/commits/content-libraries
**Dependents**: This is currently blocking work on analytics support ( https://github.com/open-craft/edx-platform/pull/26 ), which depends on these changes.
**Sandbox URL**: http://sandbox3.opencraft.com:18010/   (LMS at http://sandbox3.opencraft.com/ but shouldn't be needed)
**Internal code reviews PRs**: https://github.com/open-craft/edx-platform/pull/6 https://github.com/open-craft/edx-platform/pull/25
**Partner information**: 3rd party-hosted open edX instance, for an edX solutions client.

**Author notes**:
* Capa modules unfortunately store some of their *content* ("markdown") as `Scope.settings`, which throws a wrench into the works. For this PR I have added a capa-specific workaround (20d07d210074081a038fec781e484539f1022b31).
* There is currently no way to indicate in the studio UI which fields are Scope.settings (can be overridden) and which are Scope.content (should not be overridden as changes will be lost whenever the block is refreshed with the latest version from the library). Implementing such a notice to the user would be a complex task and is outside the scope of this PR.

bradenmacdonald commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2045,6 +2044,157 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+
+        :param user_id: The user who will get credit for making this change.
+
+        :param copy_children: If true, all descendants of each XBlock will be copied recursively.
```
Yeah, I will just remove that. It is always True.

bradenmacdonald commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2045,6 +2044,157 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+                if course.branch is None:
+                    raise ItemNotFoundError("branch is required for all source keys when using inherit_copy")
+                if course not in source_structures:
+                    with self.bulk_operations(course):
```
I just tried moving this up like you suggested but it doesn't seem to work without a valid course key. Also, 99% of the time, all of the source_keys will be coming from just one library/course, so it's only going to be calling `_lookup_course` once anyways. Thus I'm going to leave it for now - let me know if you want me to revisit.

bradenmacdonald commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2045,6 +2044,157 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+            src_course_key = usage_key.course_key.for_version(None)
+            block_key = BlockKey(usage_key.block_type, usage_key.block_id)
+            source_structure = source_structures.get(src_course_key, [])
+            if block_key not in source_structure['blocks']:
```
It's improbable, but it's not impossible - because `source_structures` is only generated from the top-level usage keys, but this method recurses to those block's children. If one of those blocks or any of its descendants has an invalid "children" setting that refers to a block missing from that course/library's structure, this could occur. That said, the only way that would happen is if at some point there was an error during some modulestore operation, resulting in the structure having an inconsistent state. (Something that has happened to me while coding changes to split, but shouldn't happen in production.)

bradenmacdonald commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py:
```diff
@@ -93,6 +93,14 @@ def _auto_publish_no_children(self, location, category, user_id, **kwargs):
+        """
+        source_keys = [self._map_revision_to_branch(key) for key in source_keys]
+        dest_key = self._map_revision_to_branch(dest_key)
+        return super(DraftVersioningModuleStore, self).inherit_copy(source_keys, dest_key, user_id, copy_children)
```
Good catch. How's this? 48b45917ce83788a4c4d30b250f96381974b09c5

bradenmacdonald commented:
Test failures [are unrelated](https://github.com/edx/edx-platform/pull/6432). Just waiting for further feedback/review.

explorerleslie commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2045,6 +2044,157 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+            # capa modules also use a custom editor that always saves their markdown field to the metadata, even if it hasn't changed, which breaks our override system.
+            # So until capa modules are fixed, we special-case them and remove their markdown fields, forcing the inherited version to use XML only.
+            if usage_key.block_type == 'problem' and 'markdown' in new_block_info['defaults']:
+                del new_block_info['defaults']['markdown']
```
thanks for the mention, @dmitchell. What are "borrowed definitions"? Do you mean, a course author using a problem that came from a library can only use the "advanced editor" option of the problem, not the default markdown editor? @cahrens can you please review @dmitchell 's comment above?

dmitchell replied:
When a course uses the content library, the xblocks (problems, html, verticals, etc) from the library become virtual parts of the course. Problems have 2 versions of the content: the human readable markdown and the xml generated from that. The problems pulled in from the content library will only have the xml; so, the course authors won't be able to use the markdown editor on those problems. However, most of this design assumes that the course authors are never editing anything they pull in from libraries (except by editing the library); so, that may not be a big deal.

bradenmacdonald replied:
@cahrens Good observation. The only reason you're seeing the advanced editor on some problems and not on others is that some of the problems were edited with an earlier version of this branch. That created an override of the "markdown" field for that problem in the course, and it will persist since it's technically a Scope.settings field.

If you delete all the blocks and then refresh the library, that problem should go away, and it should not occur on any blocks created/edited after this PR is merged.

cpennington replied:
@cahrens @explorerleslie: What's the likelyhood that we'd be able to get time to switch `markdown` to `Scope.content` at some point?

cahrens replied:
I'm not sure why markdown was ever Scope.settings...

cpennington replied:
Probably because studio only provided an editor for Scope.settings values.

e-kolpakov commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2071,6 +2070,154 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+
+            # Set of all descendent block IDs of dest_usage that are to be replaced:
+            block_key = BlockKey(dest_usage.block_type, dest_usage.block_id)
+            orig_descendants = set([block for block in self.descendants(dest_structure['blocks'], block_key, depth=None, descendent_map={})])
```
@bradenmacdonald most likely equal to 

    list(self.descendants(dest_structure['blocks'], block_key, depth=None, descendent_map={}))

e-kolpakov commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2071,6 +2070,154 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+            dest_info['edit_info']['previous_version'] = dest_info['edit_info']['update_version']
+            dest_info['edit_info']['update_version'] = old_dest_structure_version
+            dest_info['edit_info']['edited_by'] = user_id
+            dest_info['edit_info']['edited_on'] = datetime.datetime.now(UTC)
```
@bradenmacdonald this `some_info['edit_info'][...] = ...` is used quite often, so it might make sense encapsulating it into something like

    def _set_info_items(info, **kwargs):
        for key, value in kwargs.items():
            info['edit_info'][key] = value

And use like

    self._set_info_items(dest_info, 
        previous_version=dest_info['edit_info']['update_version'], 
        update_version=old_dest_structure_version,
        ...
    )

cpennington replied:
I would rather turn the structure object into an actual python object (perhaps by doing a wholesale conversion to a ObjectMapper over mongo, such as mongoengine).

bradenmacdonald replied:
Either way, that pattern is used throughout this file so I'd prefer to see it done as a separate refactor.

antoviaque commented:
Sandbox updated with the latest code: http://sandbox3.opencraft.com:18010/

cahrens commented:
I created https://openedx.atlassian.net/browse/TNL-1109 for the TNL review.

antoviaque commented:
@cahrens For the second reviewer for the common modulestore code, who do you think would be good for this?

cahrens commented:
Someone on Solutions or another Platform member? We generally want 2 reviewers on code if it is all complicated.

antoviaque commented:
@cahrens OK, thanks - on Solutions Chris and Eugeny had done reviews, but they are both not very familiar with the code, so it might make sense to get someone else from platform. Maybe @cpennington or @doctoryes ?

bradenmacdonald commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -552,7 +552,10 @@ def _duplicate_item(parent_usage_key, duplicate_source_usage_key, user, display_
 
         # Update the display name to indicate this is a duplicate (unless display name provided).
-        duplicate_metadata = own_metadata(source_item)
+        duplicate_metadata = {}
```
It wasn't an issue before because most XBlocks have simple types for their settings. LibraryContent blocks use a unique `LibraryList` value type which contains `LibraryVersionReference` objects which contain `LibraryLocator` and `ObjectId` instances.

Three representations are:
* JSON String: `'[["library-v1:ProblemX+PR0B", "5436ffec56c02c13806a4c1b"], ...]'` (`unicode` type)
* Intermediate: `[["library-v1:ProblemX+PR0B", "5436ffec56c02c13806a4c1b"], ...]`(`list` type)
* Native: `[(LibraryLocator(org="ProblemX", library="PR0B"), ObjectId("5436ffec56c02c13806a4c1b")), ...]` (`LibraryList` type)

Unfortunately it's easy to mix and match the intermediate and the native types for most XBlocks - but for this one in particular it breaks things. Part of the issue is the API design of `from_json` and `to_json` - if they always took and returned *strings*, there would be no confusion, but since they take and return intermediate python types which resemble native types 99% of the time, we get this situation. I guess that the intermediate types should never really be used outside of the XBlock internal/runtime/fieldstore context but it appears that they are.

dmitchell replied:
Yeah, the conversion to-from json is an ongoing confusion and problem which I don't think we've systematically encapsulated at the right places. There are probably other places which have similar issues. imho, I wish field data used python native types and not json and the only conversion to/from json would happen only at serialization to external services/clients.

cahrens replied:
Thank you for the detailed explanation-- that helped. Please just include a brief comment in the code stating that we cannot use own_metadata because of the intermediate representation issue (which would break LibraryList, for example).

dmitchell commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -306,3 +306,57 @@ def test_change_after_first_sync(self):
+
+        # Make sure that the new child of the LibraryContent block
+        # shares its definition with block1
+        block1_course = modulestore().get_item(lc_block1.children[0])
```
Calling this `_course` confused me for a sec as I assumed it was the course root. idk if it was just me this am.

bradenmacdonald replied:
`block1_in_course` perhaps?

bradenmacdonald replied:
I've renamed it to `problem_in_course` in the updated tests which I'm about to push.

dmitchell commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -306,3 +306,57 @@ def test_change_after_first_sync(self):
+
+        # Now make sure the override persists even when blocks are refreshed.
+        # Force a new library version by adding another block to the library
+        ItemFactory.create(category="problem", parent_location=self.library.location, user_id=self.user.id, publish_item=False)
```
Should you include a test for the override after refresh? That is, change the library's default and refresh the courses use of it and then do the same fetch assertEqual tests you have here?

dmitchell commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -306,3 +306,57 @@ def test_change_after_first_sync(self):
+        self.assertEqual(block1_course.display_name, original_display_name)
+
+        # Change a settings field on lc_block1
+        block1_course.display_name = "NEW"
```
My head's a big foggy today w/ this cold. Given the special purpose handling of display_name in _duplicate_item, I'm wondering whether you should test using a different settings field to ensure all settings handle overrides correctly?

bradenmacdonald replied:
Good idea, I have expanded and refactored the tests and now they will check with "weight" as well as display_name.

dmitchell commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -301,14 +301,11 @@ def refresh_children(self, request, suffix, update_db=True):  # pylint: disable=
         the version number of the libraries used, so we easily determine if
         this block is up to date or not.
-
-        If update_db is True (default), this will explicitly persist the changes
```
I think that's because I pointed out to him that there was no obv use case nor test for `update_db=False`

dmitchell commented on common/lib/xmodule/xmodule/library_tools.py:
```diff
@@ -55,74 +53,18 @@ def update_children(self, dest_block, user_id, update_db=True):
-        root_children = []
+        new_libraries = []
+        source_blocks = []
+        for library_key, dummy in dest_block.source_libraries:
```
I believe our standard for unused vars is `__` but I don't care as long as you're not getting pylint errors. (I think pylint ignores vars starting w/ `_`)

cpennington commented on cms/djangoapps/contentstore/tests/test_libraries.py:
```diff
@@ -306,3 +306,57 @@ def test_change_after_first_sync(self):
+        # Make sure that the new child of the LibraryContent block
+        # shares its definition with block1
+        block1_course = modulestore().get_item(lc_block1.children[0])
+        self.assertEqual(block1_course.definition_locator.definition_id, def_id1)
```
`definition_locator` is a split-mongo specific implementation detail. You'd be better off comparing `.scope_ids.def_id`.

bradenmacdonald replied:
@cpennington In this case, I am trying to compare the `definition_id` specifically (a mongo ObjectId), and it is a test that is specific to split mongo. I want to make sure that when a block is copied from a library that the definition object in Split's definitions store is re-used. The test wouldn't make sense on other modulestores.

I tried what you suggested but that's a totally different thing:
```
AssertionError: BlockUsageLocator(CourseLocator('org.37', 'course_37', 'Run_37', None, None), u'problem', u'7b2359901823908cc241') != LibraryUsageLocator(LibraryLocator(u'org', u'lib', None, None), 'problem', 'A_Problem_Block')
```

cpennington replied:
Huh. I'm surprised that you got `UsageLocator`s when asking for `.def_id`. I would have expected `DefinitionLocator`s.

If the test is explicitly split_mongo specific, then that's probably ok.

dmitchell commented:
:+1:

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2071,6 +2070,154 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
             self._update_head(destination_course, index_entry, destination_course.branch, destination_structure['_id'])
 
+    @contract(source_keys="list(BlockUsageLocator)", dest_usage=BlockUsageLocator)
+    def inherit_copy(self, source_keys, dest_usage, user_id):
```
Using `inherit` here worries me, since there's a risk of conflating it with `InheritanceMixin` and other similar code.

bradenmacdonald replied:
What would be better? `copy_from_template`?

cpennington replied:
Yeah, that would be better.

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split.py:
```diff
@@ -2071,6 +2070,154 @@ def copy(self, user_id, source_course, destination_course, subtree_list=None, bl
+                unicode(src_course_key).encode("utf-8"),
+                block_key.id,
+                new_parent_block_key.id,
+            )
```
Seeing things like this makes me really wish that we'd just gone for integer keys for identity. This is going to produce a monster of a key.

cpennington replied:
Oh. I just noticed the next line where you hash the whole thing.

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_mongo_kvs.py:
```diff
@@ -19,17 +19,19 @@ class SplitMongoKVS(InheritanceKeyValueStore):
 
         :param definition: either a lazyloader or definition id for the definition
         :param initial_values: a dictionary of the locally set values
+        :param default_values: any Scope.settings fields that are set locally (inherited from another block)
```
Again, I'm concerned about using the word `inherited` here.

cpennington commented on common/test/acceptance/tests/studio/test_studio_library_container.py:
```diff
@@ -162,3 +162,54 @@ def test_out_of_date_message(self):
+        Then I can see that the overrided version of the setting is preserved.
+
+        When I click to edit the block and reset the setting,
+        then I can see that the setting's field defaults back to the library version.
```
This test format description seems like overkill for bok_choy tests.

cahrens replied:
As a best practice we do want BDD test descriptions for bok choy tests. The empty rows are not required, but are here for ease of reading (I assume). I think that part of the issue here is that this test is doing a ton of different things, when really it could be broken up into shorter, independent tests.

bradenmacdonald replied:
Note: I made this one long test because (A) bok choy tests are really slow and I assume one combined test is faster than many small tests with repeated setup, and (B) we were told to aim for about one bok choy test per story with this libraries work for the same reason.

bradenmacdonald commented:
Pushed some fixes that address almost all comments so far - just waiting for advice on a better name for `inherit_copy`.

cahrens commented on common/lib/xmodule/xmodule/modulestore/inheritance.py:
```diff
@@ -211,8 +211,8 @@ def inherit_metadata(descriptor, inherited_data):
-    Return a dictionary that contains only non-inherited field keys,
-    mapped to their serialized values
+    Return a JSON-friendly dictionary that contains only non-inherited field
+    keys, mapped to their serialized values
```
Some more details about this "intermediate" form would be nice, since you understand it so well now. :)

cahrens commented:
:+1: From TNL perspective, once there is a resolution about "inherited" and "inherit_copy".

bradenmacdonald commented:
@cpennington Thanks. I renamed inherit_copy to `copy_from_template` in c005bdb5ee453e29d2fd2f2daa6c9e5f04e13c16, then I squashed this PR's commits. Is this good to go now, assuming the build passes?

@cahrens Thanks for the review and the thumbs up!

bradenmacdonald commented:
Just found a bug so I pushed a fix and a test for it ("test_copy_from_template_publish"). While doing that I moved some of the new tests to their own file as Christina had suggested.

cpennington commented:
:+1:

bradenmacdonald commented:
Thanks everyone! Merging now.
